Marlin/src/gcode/control/M42.cpp

```diff
@@ -28,6 +28,11 @@
  * M42: Change pin status via GCode
  *
  *  P<pin>  Pin number (LED if omitted)
+ *            For LPC1768 use M42 P1.20 S255 if wanting to set P1_20 to logic 1
+ *              NOTE - Repetier Host truncates trailing zeros on a decimal when
+ *                     sending commands so typing M42 P1.20 S255 results in 
+ *                     M42 P1.2 S255 being sent.  Pronterface doesn't have this issue.
+ *                     
```

Comment:
We don't need to keep `M42` in this form for MCUs that use the Pr_pp pin numbering.
I suggest `M42 R2 P20 S255` where `R` is the port, and `P` is the port pin.

Reply:
Thanks

Reply:
I tried that format and it didn't work.

Do you want me to change M42?

Reply:
We could change the Pr_pp format to require a leading 0 on the rr part. This would force you to type `M42 P1.02` and then we could interpret `P1.2` or `P1.20` as the same. It's not really ideal, but it would work.

Reply:
The problem is the decimal '.'.
RH is checking all G0/G1 commands coordinates against its internal soft endstops. For that it's obviously reading the command line,, converting numbers with decimal point to floats and converting them back to strings when sent to Marlin.
How about P1_20 or P1,20?

Reply:
P1_20 and P1,20 do not work.

Repetier Host is a limiting factor.  P1_18, P1,18, PP1_18 all get sent as P0.

Pronterface sends them as is.

How about doing a separate PR for M42 once it's been decided how we want it to work?

Reply:
`M42 P1.02`

print (P)

`1.01999999999`

Reply:
> I tried that format and it didn't work.

Naive question… Did you implement the changes to `M42` before trying that?

Reply:
I did a little playing with the Repetier Host ini file along with a couple of minor code changes and got "M42 L1_20 S50" to modify P1_20 properly.

The Repetier Host change was to add an optional file name to the M42 command.  Unfortunately Repetier Host still translated P1_20 to P0 when it sent the command.

Scott's original suggestion is looking better & better. Rx and Py where x is port and y is pin.  We could change the LPC1768's PARSED_PIN_INDEX function to do the translation.

Reply:
I just did a commit that implements Scott's suggestion.  Let me know if we should keep it or not.

Reply:
Yes, P1.02 works but P1.20 doesn't work in Repetier Host.

Reply:
I just noticed that M43 has S and E for start pin number and end pin number.  That means the R & P idea won't work for M43.

Reply:
For a range of pins, I agree `S` and `E` are perhaps too English-centric. Since `M43` is not canonical we can mash it up any way we like. So, I suggest we use `I`,`J` for start-end pins. And, if we are _truly determined_ to avoid any single-value pin numbering for ARM32 then `K` and `L` for start-end ports. `M43 K1 I20 J22` would specify the range P1_20 to P1_22.

Reply:
Fool idea I was discussing with an engineer here… Create an "emulated mapping" that mimics some board with a lot of well-defined pins (such as an Azteeg X3) — applying a "fastio pin numbering" scheme that conceals the real pin numbers. Then, create an "extended pin mapping" that exposes the rest of the pins.

Obviously not ideal, and it would make it hard to correlate pins to the datasheets. But, since we are basically creating something like an emulation of Marlin/AVR it almost makes sense. Anyway, feel free to scoff at this concept. I'm sure that as I delve more into these architectures I'll realize that pin correlations aren't that simple.